ic-identity-hsm/src/hsm.rs

```diff
@@ -0,0 +1,358 @@
+        let key_id = str_to_key_id(key_id)?;
+        let public_key = get_der_encoded_public_key(&ctx, session_handle, &key_id)?;
+
+        Ok(HardwareIdentity {
+            key_id,
+            ctx,
+            session_handle,
+            logged_in,
+            public_key,
+        })
+    }
+}
+
+impl Identity for HardwareIdentity {
+    fn sender(&self) -> Result<Principal, String> {
+        Ok(Principal::self_authenticating(&self.public_key))
+    }
+    fn sign(&self, msg: &[u8], _principal: &Principal) -> Result<Signature, String> {
+        let hash = hash_message(msg);
+        let signature = self.sign_hash(&hash)?;
+
+        Ok(Signature {
+            public_key: self.public_key.clone(),
+            signature,
+        })
+    }
+}
+
+// We open a session for the duration of the lifetime of the HardwareIdentity.
+fn open_session(
+    ctx: &Ctx,
+    slot_id: CK_SLOT_ID,
+) -> Result<CK_SESSION_HANDLE, HardwareIdentityError> {
+    let flags = CKF_SERIAL_SESSION;
+    let application = None;
+    let notify = None;
+    let session_handle = ctx.open_session(slot_id, flags, application, notify)?;
+    Ok(session_handle)
+}
+
+// We might need to log in.  This requires the PIN.
+fn login_if_required(
+    ctx: &Ctx,
+    session_handle: CK_SESSION_HANDLE,
+    pin: &str,
+    slot_id: CK_SLOT_ID,
+) -> Result<bool, HardwareIdentityError> {
+    let token_info = ctx.get_token_info(slot_id)?;
+    let login_required = token_info.flags & CKF_LOGIN_REQUIRED != 0;
+
+    if login_required {
+        ctx.login(session_handle, CKU_USER, Some(&pin))?;
+    }
+    Ok(login_required)
+}
+
+// Return the DER-encoded public key in the expected format.
+// We also validate that it's an ECDSA key on the correct curve.
+fn get_der_encoded_public_key(
+    ctx: &Ctx,
+    session_handle: CK_SESSION_HANDLE,
+    key_id: &KeyId,
+) -> Result<DerPublicKeyVec, HardwareIdentityError> {
+    let object_handle = get_public_key_handle(&ctx, session_handle, key_id)?;
+
+    validate_key_type(ctx, session_handle, object_handle)?;
+    validate_ec_params(ctx, session_handle, object_handle)?;
+
+    let ec_point = get_ec_point(&ctx, session_handle, object_handle)?;
+
+    let oid_ecdsa = oid!(1, 2, 840, 10045, 2, 1);
+    let oid_curve_secp256r1 = oid!(1, 2, 840, 10045, 3, 1, 7);
+    let ec_param = Sequence(
+        0,
+        vec![
+            ObjectIdentifier(0, oid_ecdsa),
+            ObjectIdentifier(0, oid_curve_secp256r1),
+        ],
+    );
+    let ec_point = BitString(0, ec_point.len() * 8, ec_point);
+    let public_key = Sequence(0, vec![ec_param, ec_point]);
+    let der = to_der(&public_key)?;
+    Ok(der)
+}
+
+// Ensure that the key type is ECDSA.
+fn validate_key_type(
+    ctx: &Ctx,
+    session_handle: CK_SESSION_HANDLE,
+    object_handle: CK_OBJECT_HANDLE,
+) -> Result<(), HardwareIdentityError> {
+    // This value will be mutated.  `let mut` emits a warning, though.
```

Comment:
is it mutated because of ffi? Could you add a note for "why" it's mutated and maybe wrap this in unsafe to make it double clear (this is super surprising to the reader and may lead to bugs).

Reply:
I don't know what `ffi` is.

Reply:
https://en.wikipedia.org/wiki/Foreign_function_interface

Basically bindings between libraries built in different languages. I think the whole pkcs11 library uses FFI to link against a library that's really in C, right?

Reply:
Maybe that is how it gets around it!